<a href="https://colab.research.google.com/github/bngrossmann/new-york-crime-analysis/blob/main/New_York_Crime_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 2

* Benjamin Grossmann

# Part 1


Your second project is going to have a lot more freedom than your first project. This is because we want you to have a project in your portfolio that interests you or relates to the industry you would like to work in.

Your task for this week is to propose two possible datasets you would like to work with for Project 2.  

You will choose your first choice data set, and a backup data set in case the first proposed data set is not approved.  

This data can be from any source and can be on any topic with these limitations:

the data must be available for use (it is your responsibility to ensure that the license states that you are able to use it.)
the data must be appropriate for a professional environment
the data must NOT contain personal information
the data must NOT be a dataset used for any assignment, lecture, or task from the course
Make sure you select a dataset that will be reasonable to work with in the amount of time we have left. Think about what questions you could reasonably answer with the dataset you select. 

You must propose two datasets that each have a supervised learning component. You may choose a regression or classification problem for each proposed data set.

**For this task:**

Create a Colab notebook where you have uploaded and shown the .head() of each of your data sets.  For each of the proposed datasets, answer the following questions:

## **Dataset #1**


In [390]:
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [391]:
# Load your data & imports
import numpy as np
import pandas as pd

In [392]:
df = pd.read_csv("/content/drive/MyDrive/Coding Dojo/Data Folder/NYPD_Complaint_Data_Current__Year_To_Date_.csv")

In [393]:
# .head() for your dataframe
df.head()

CMPLNT_NUM  ADDR_PCT_CD BORO_NM CMPLNT_FR_DT CMPLNT_FR_TM CMPLNT_TO_DT  \
0   903695881           69     NaN   12/17/2021     22:13:00          NaN   
1   400462399          113     NaN   12/17/2021     06:21:00          NaN   
2   587910690          113     NaN   12/13/2021     20:05:00          NaN   
3   186105368           52   BRONX   12/07/2021     22:49:00          NaN   
4   185325394          113     NaN   12/06/2021     17:25:00          NaN   

  CMPLNT_TO_TM CRM_ATPT_CPTD_CD HADEVELOPT  HOUSING_PSA  ...  \
0          NaN        COMPLETED        NaN          NaN  ...   
1          NaN        COMPLETED        NaN          NaN  ...   
2          NaN        COMPLETED        NaN          NaN  ...   
3          NaN        COMPLETED        NaN          NaN  ...   
4          NaN        COMPLETED        NaN          NaN  ...   

   TRANSIT_DISTRICT VIC_AGE_GROUP  VIC_RACE VIC_SEX X_COORD_CD Y_COORD_CD  \
0               NaN         25-44     BLACK       M    1011203     174515   
1               NaN         25-44     BLACK       F    1043252     187998   
2               NaN         25-44     BLACK       M    1042087     190443   
3               NaN       UNKNOWN   UNKNOWN       E    1017088     260895   
4               NaN         25-44     BLACK       M    1046176     193100   

    Latitude  Longitude                                   Lat_Lon  \
0  40.645647 -73.902876   (40.64564719600002, -73.90287588699994)   
1  40.682499 -73.787269  (40.682499421000045, -73.78726915499993)   
2  40.689218 -73.791449  (40.689218058000044, -73.79144856999993)   
3  40.882718 -73.881247   (40.88271780700006, -73.88124724999993)   
4  40.696483 -73.776680   (40.69648322200004, -73.77667979199998)   

                        New Georeferenced Column  
0   POINT (-73.90287588699994 40.64564719600002)  
1  POINT (-73.78726915499993 40.682499421000045)  
2  POINT (-73.79144856999993 40.689218058000044)  
3   POINT (-73.88124724999993 40.88271780700006)  
4   POINT (-73.77667979199998 40.69648322200004)  

[5 rows x 36 columns]

In [394]:
df.shape

(449506, 36)

## Questions

### 1) source of your data

The data comes from the New York City Police Department.

https://data.cityofnewyork.us/Public-Safety/NYPD-Complaint-Data-Current-Year-To-Date-/5uac-w243


### 2) Brief description of data

The data includes report information about crimes reported to the NYPD. (Location: Coordinates and Descriptive; Day and Time; Suspect Profile Information; Victim Profile Information; Weather; and several others)

### 3) What is the target?

The target is `"LAW_CAT_CD"`, which is the level of offense (felony, misdemeanor, violation).

###  4) Is this a classification or regression problem?

This will be a classification problem.

### 5) How many features?
Although the dataset has 35 features after being separated from the target, many of them are unnecessary and should be removed before any modeling is done.

For example, there are multiple location features. Two pair of which express the same coordinate information in two different coordiante systems.

There are also several features that are codes which ways of encoding report information and should not be used. They do not predict the target. They were explictly created by the reporting system after the crime was reported and cannot be used to predict future events (they postdict).

A thourough examination of the data will be required to determine all the features that should be removed before modeling.


### 6) How many rows of data.

The dataset has 449,506 rows of data

### 7) What, if any, challenges do your foresee in cleaning, exploring, or modeling with this dataset?

There are several columns which have a lot of missing values. Some missing values may be acceptable (`"PARKS_NM"` has no value if the occurance did not happen in a park), and can be accomadated.

And as mentioned above in question 5, several of the columns should not be used because they may have values that are determined by the target values. Determining which ones may take some research.

Some columns (such as some of the description columns) have multiple values within a single cell. It may be necessary to break them apart into their component values.

#Data Dictionary

Column Name|Description|Type
-|-|-
CMPLNT_NUM|Randomly generated persistent ID for each complaint|Number
ADDR_PCT_CD|The precinct in which the incident occurred|Number
BORO_NM|The name of the borough in which the incident occurred|Plain Text
CMPLNT_FR_DT|Exact date of occurrence for the reported event (or starting date of occurrence, if CMPLNT_TO_DT exists)|Date & Time
CMPLNT_FR_TM|Exact time of occurrence for the reported event (or starting time of occurrence, if CMPLNT_TO_TM exists)|Plain Text
CMPLNT_TO_DT|Ending date of occurrence for the reported event, if exact time of occurrence is unknown|Date & Time
CMPLNT_TO_TM|Ending time of occurrence for the reported event, if exact time of occurrence is unknown|Plain Text
CRM_ATPT_CPTD_CD|Indicator of whether crime was successfully completed or attempted, but failed or was interrupted prematurely|Plain Text
HADEVELOPT|Name of NYCHA housing development of occurrence, if applicable|Plain Text
HOUSING_PSA|Development Level Code|Number
JURISDICTION_CODE|Jurisdiction responsible for incident. Either internal, like Police(0), Transit(1), and Housing(2); or external(3), like Correction, Port Authority, etc.|Number
JURIS_DESC|Description of the jurisdiction code|Plain Text
KY_CD|Three digit offense classification code|Number
LAW_CAT_CD|Level of offense: felony, misdemeanor, violation|Plain Text
LOC_OF_OCCUR_DESC|Specific location of occurrence in or around the premises; inside, opposite of, front of, rear of|Plain Text
OFNS_DESC|Description of offense corresponding with key code|Plain Text
PARKS_NM|Name of NYC park, playground or greenspace of occurrence, if applicable (state parks are not included)|Plain Text
PATROL_BORO|The name of the patrol borough in which the incident occurred|Plain Text
PD_CD|Three digit internal classification code (more granular than Key Code)|Number	
PD_DESC|Description of internal classification corresponding with PD code (more granular than Offense Description)|Plain Text
PREM_TYP_DESC|Specific description of premises; grocery store, residence, street, etc.|Plain Text
RPT_DT|Date event was reported to police|Date & Time
STATION_NAME|Transit station name|Plain Text
SUSP_AGE_GROUP|Suspect’s Age Group|Plain Text
SUSP_RACE|Suspect’s Race Description|Plain Text
SUSP_SEX|Suspect’s Sex Description|Plain Text
TRANSIT_DISTRICT|Transit district in which the offense occurred.|Number	
VIC_AGE_GROUP|Victim’s Age Group|Plain Text
VIC_RACE|Victim’s Race Description|Plain Text
VIC_SEX|Victim’s Sex Description|Plain Text
X_COORD_CD|X-coordinate for New York State Plane Coordinate System, Long Island Zone, NAD 83, units feet (FIPS 3104)|Number
Y_COORD_CD|Y-coordinate for New York State Plane Coordinate System, Long Island Zone, NAD 83, units feet (FIPS 3104)|Number
Latitude|Midblock Latitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326)|Number
Longitude|Midblock Longitude coordinate for Global Coordinate System, WGS 1984, decimal degrees (EPSG 4326)|Number
Lat_Lon||Location
New Georeferenced Column||Point

# 1 . Load Data

In [395]:
# from google.colab import drive
# drive.mount("/content/drive")

In [396]:
# import numpy as np
# import pandas as pd

In [397]:
# df = pd.read_csv("/content/drive/MyDrive/Coding Dojo/Data Folder/NYPD_Complaint_Data_Current__Year_To_Date_.csv")

In [398]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 449506 entries, 0 to 449505
Data columns (total 36 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   CMPLNT_NUM                449506 non-null  int64  
 1   ADDR_PCT_CD               449506 non-null  int64  
 2   BORO_NM                   448355 non-null  object 
 3   CMPLNT_FR_DT              449506 non-null  object 
 4   CMPLNT_FR_TM              449506 non-null  object 
 5   CMPLNT_TO_DT              409416 non-null  object 
 6   CMPLNT_TO_TM              409568 non-null  object 
 7   CRM_ATPT_CPTD_CD          449345 non-null  object 
 8   HADEVELOPT                1577 non-null    object 
 9   HOUSING_PSA               32583 non-null   float64
 10  JURISDICTION_CODE         449023 non-null  float64
 11  JURIS_DESC                449506 non-null  object 
 12  KY_CD                     449506 non-null  int64  
 13  LAW_CAT_CD                449506 non-null  o

In [399]:
# Function for showing unique values and missing rows from columns of a dataframe
def show_column_values(dataframe,column_list=[]):
    for col in column_list:
        print(f"{col:>18} : {dataframe[col].nunique():>5} : {dataframe[col].value_counts().sort_index().index.to_list()}")
        print(f"{':':>20} {dataframe.loc[ df[col].isna() , col].duplicated().sum():>5} : Rows Missing Data\n")

## 1 . 2 . Identify the Target

The target value for this analysis is the level of the offense.
* `"LAW_CAT_CD"`
    > The level of the offense: felony, misdemeanor, or violation.

In [400]:
# for col in ["LAW_CAT_CD"]:
#     print(f"{col:>14} : {df[col].nunique():>5} : {df[col].value_counts().sort_index().index.to_list()}")
#     print(f"{':':>16} {df.loc[ df[col].isna() , col].duplicated().sum():>5} : Rows Missing Data\n")
show_column_values(df,["LAW_CAT_CD"])

        LAW_CAT_CD :     3 : ['FELONY', 'MISDEMEANOR', 'VIOLATION']
                   :     0 : Rows Missing Data



# 2 . Uniqueness and Duplication

Each data row is suppose to represent a single unique complaint identified in the `"CMPLNT__NUM"` column. So there should be no duplicates.

In [401]:
print(f"{df.shape[0]} Rows in the data table")
print(f"{df['CMPLNT_NUM'].nunique()} Unique Entries in the complaint number column")
print(f"{df.duplicated().sum():>6} Duplicated table rows")

449506 Rows in the data table
449506 Unique Entries in the complaint number column
     0 Duplicated table rows


Verified that no rows are duplicated.

Since the values for `"CMPLNT_NUM"` are unique integers for each row, they can be used as the row index.

In [402]:
df.set_index("CMPLNT_NUM",inplace=True);

Now that the unique complaint number has been changed to the index, I can check again for duplicate rows. These new duplicates will indicate that there were probably mulitple suspects or mutiple victims at each incident, with each addtional suspect or victim contributing to new rows in the data table.

In [403]:
print(f"{df.duplicated().sum()} Duplicated table rows")

1576 Duplicated table rows


Now I must make a choice of whether I will drop these new duplicates. I can either drop them and only consider each group of new duplicates as a single event. Or I can keep them and consider each new duplicate a separate event.

From a resource perspective, I will consider each new duplicate a separate event. Although these new duplicates happen in groups, it may require more resources to respond to a single group of events.

This choice may effect the results based on how often these groups of event occur.

In [404]:
print(f"New duplicates account for {df.duplicated().sum() / df.shape[0]*100:.2f}% of all complaints.")

New duplicates account for 0.35% of all complaints.


Since the new duplicates only account for just over a third of one percent, they are not very common (less than four out of every thousand).

Further justification for keeping these new duplicate can be found by considering who may be invovled. These duplicates are those where the suspects and victims necessarily share demographic characteristics. If a single group of events had suspects and victims with different demographic characteristics, then data rows conresponding the these different individuals would not be duplicates.

At this stage in the analysis, I am not yet prepared to determine any group events that have non-duplicated rows of data. After I do so, it will be possible to do comparative analysis to see if grouping simultaneous events as a single occurance has any detectable consequences. There may be some applications where such grouping is a better measure than haveing them treated as separate these events.

# 3 . Eliminating Unnecessary Columns

In building a model that can help me understand where different kinds of crimes are happening, I need to separate the information that can be used to help allocate resources from the information that which cannot.

Time and location are quantities that must occur for any complaint, and exist even if there is no complaint. These are also important criteria to use when allocating resorces. Without these pieces of information there is no framework for understanding the layout of the city.

Since I will be dropping columns several times, I will make a function to assist me and report how many columns remain.


In [405]:
# Function for dropping a list of columns from a dataframe and reporting it
def drop_columns(dataframe, column_list=[]):
    print(f"{dataframe.shape[1]} columns before the drop.")
    dataframe.drop(columns=column_list, inplace=True)
    print(f"{dataframe.shape[1]} columns after the drop.")

## 3 . 1 . Suspect and Victim Demographics

There are only three characteristics in the data set given for suspect and victim demographics: sex, age, and race.
* `"SUSP_SEX"`, `"VIC_SEX"`
> Sex of the suspect or victim
* `"SUSP_RACE"`, `"VIC_RACE"`
> Race of the suspect or victim
* `"SUSP_AGE_GROUP"`, `"VIC_AGE_GROUP"`
> Age group of the suspect or victim

I can look at how these values are recorded. I will display the values in ascending order (alphabetically and increasing numerically) and how much data is missing.

In [406]:
# for col in ["SUSP_SEX","SUSP_RACE","SUSP_AGE_GROUP","VIC_SEX","VIC_RACE","VIC_AGE_GROUP"]:
#     print(f"{col:>14} : {df[col].nunique():>5} : {df[col].value_counts().sort_index().index.to_list()}")
#     print(f"{':':>16} {df.loc[ df[col].isna() , col].duplicated().sum():>5} : Rows Missing Data\n")

show_column_values(df,["SUSP_SEX","SUSP_RACE","SUSP_AGE_GROUP","VIC_SEX","VIC_RACE","VIC_AGE_GROUP"])

          SUSP_SEX :     3 : ['F', 'M', 'U']
                   : 88133 : Rows Missing Data

         SUSP_RACE :     7 : ['AMERICAN INDIAN/ALASKAN NATIVE', 'ASIAN / PACIFIC ISLANDER', 'BLACK', 'BLACK HISPANIC', 'UNKNOWN', 'WHITE', 'WHITE HISPANIC']
                   : 88133 : Rows Missing Data

    SUSP_AGE_GROUP :    26 : ['-33', '-60', '-69', '-941', '-946', '-947', '-955', '-966', '-969', '-971', '-973', '-975', '1017', '1032', '1056', '18-24', '2021', '25-44', '45-64', '65+', '938', '940', '942', '953', '<18', 'UNKNOWN']
                   : 88133 : Rows Missing Data

           VIC_SEX :     4 : ['D', 'E', 'F', 'M']
                   :     0 : Rows Missing Data

          VIC_RACE :     7 : ['AMERICAN INDIAN/ALASKAN NATIVE', 'ASIAN / PACIFIC ISLANDER', 'BLACK', 'BLACK HISPANIC', 'UNKNOWN', 'WHITE', 'WHITE HISPANIC']
                   :    80 : Rows Missing Data

     VIC_AGE_GROUP :    21 : ['-1', '-3', '-4', '-48', '-51', '-61', '-62', '-921', '-935', '-943', '-960', '18-24',

These won't be very helpful to understand how crime is distributed as people matching these descriptions can be found throughout the city. Further, many complaints have no information about the suspect. And victims can include businesses and organizations (sex `"D"`) and the People of the State of New York  (sex `"E"`)

A quick calculation can even show how much suspect information is unavailable. (As per the NYPD Complaints Incident Level Data Footnotes, item number 2, null values should be considered as either "Unknown/Not Available/Not Reported").

In [407]:
for col,val in zip(("SUSP_SEX","SUSP_RACE","SUSP_AGE_GROUP"),("U","UNKNOWN","UNKNOWN")):
    print(f"{df.loc[ df[col]==val , col].count() /  df[col].shape[0] * 100:.2f}% Rows with {col} {val}")
    print(f"{df[col].isna().sum() /  df[col].shape[0] * 100:.2f}% Rows with {col} null")
    print(f"{(df[col].isna().sum()+df.loc[ df[col]==val , col].count()) /  df[col].shape[0] * 100:.2f}% Total Rows\n")

21.14% Rows with SUSP_SEX U
19.61% Rows with SUSP_SEX null
40.75% Total Rows

24.82% Rows with SUSP_RACE UNKNOWN
19.61% Rows with SUSP_RACE null
44.42% Total Rows

37.32% Rows with SUSP_AGE_GROUP UNKNOWN
19.61% Rows with SUSP_AGE_GROUP null
56.92% Total Rows



With this much information unavailable (over 40%), these columns would be problamatic if they were kept. So I will drop these from the working dataframe. (This is even without considering all the additional age values that don't represent real ages.)

In [408]:
drop_columns(df,["SUSP_SEX","SUSP_RACE","SUSP_AGE_GROUP","VIC_SEX","VIC_RACE","VIC_AGE_GROUP"])

35 columns before the drop.
29 columns after the drop.


## 3 . 2 . Location Data

There are several columns that descibe the location of the complaint. Some of them are too general, others are too specific, and some are redudant.



Primary Location Information
* `"ADDR_PCT_CD"` (Precinct) (KEEP)
>New York is divided into 77 precincts. This is the level that neighborhood communities and law enforcement will typically interact. Additionally, some offenses are assigned alternate locations than where they actually occured. (NYPD Complaints Incident Level Data Footnotes, items number 10 through 13, 15). This makes location resolution smaller than the precinct level less accurate.

Too General
*  `"BORO_NM"` (Borough) (DROP)
>New York City has five boroughs. Each has a significantly large population density (four of which are greater than any other city in the country). Locating where events occur by borough is too broad, so I will drop the column. If for some reason the boroughs should become an important feature to include, then it can be added back in as each precinct is within a specific borough (there is the exception of Marble Hill).

* `"PATROL_BORO"` (Patrol Borough) (DROP)
>The patrol borough value  divides three of the boroughs into a north and south region. But otherwise, the size of these patrol boroughs are still too broad.

Too Specific and Redundant

* `"Latitude"` and `"Longitude"` (DROP)
    > Geographical location given in the Global Coordinate System. If a smaller resolution than precinct should be needed, these coordinates can be used to help reconstruct the necessary information information. If doing so, it must be kept in mind that coordinates which are not relocated will be approximations to the nearest mid-block or street intersection. This falls in line with the NYPD Complaints Incident Level Data Footnotes, item 9, which states that trying to match locations to an exact address is not recommended.
* `"Lat_Lon"` and `"New Georeferenced Column"` (DROP)
    > The latitude and longitude coordinates given as an ordered pair. The new georeferenced column reverses the order of the latitude-longitude ordered pair. Because these ordered pairs contain the same information as the latitude and longitude, they provide no new information.
* `"X_COORD_CD"` and `"Y_COORD_CD"` (DROP)
    >  Geographical location given in the New York State Plane Coordinate System. These coordinates can be determined by transforming the latitude and longitude, so provide no new information.

In [409]:
drop_columns(df,["BORO_NM", "PATROL_BORO", "X_COORD_CD","Y_COORD_CD","Latitude","Longitude","Lat_Lon","New Georeferenced Column"])

29 columns before the drop.
21 columns after the drop.


## 3 . 3 . Location Descriptions

These columns describe attributes about the locations where the complaint occurred.

* `"PREM_TYP_DESC"` (Premise Type Description) (KEEP)
>This describe the type of premise where the complaint occurred. The distribution of location types may be a factor in identifying areas more susceptible to crimes as there may be an association between the type of crime and the type of location.
>
>This is not a smaller resolution than the precinct, as for most values it doesn't identify where within a precinct an event occurs. (Some exceptions can be identified, but are not common. For example, if the value is `"CEMETARY"` and there is only one cemetary in the precinct if won't be difficult to geo-locate the crime.)

* `"LOC_OF_OCCUR_DESC"` (DROP)
>This describes the specific location on the premises the crime occured. Although important for an investigation, I don't believe this is particularly helpful for understanding the distribution of crime throughout the city.
 
The values in each column are listed below.






In [410]:
# for col in ["PREM_TYP_DESC","LOC_OF_OCCUR_DESC"]:
#     print(f"{col:>17} : {df[col].nunique():>4} : {df[col].value_counts().index.to_list()}\n")

show_column_values(df,["PREM_TYP_DESC","LOC_OF_OCCUR_DESC"])

     PREM_TYP_DESC :    74 : ['ABANDONED BUILDING', 'AIRPORT TERMINAL', 'ATM', 'BANK', 'BAR/NIGHT CLUB', 'BEAUTY & NAIL SALON', 'BOOK/CARD', 'BRIDGE', 'BUS (NYC TRANSIT)', 'BUS (OTHER)', 'BUS STOP', 'BUS TERMINAL', 'CANDY STORE', 'CEMETERY', 'CHAIN STORE', 'CHECK CASHING BUSINESS', 'CHURCH', 'CLOTHING/BOUTIQUE', 'COMMERCIAL BUILDING', 'CONSTRUCTION SITE', 'DAYCARE FACILITY', 'DEPARTMENT STORE', 'DOCTOR/DENTIST OFFICE', 'DRUG STORE', 'DRY CLEANER/LAUNDRY', 'FACTORY/WAREHOUSE', 'FAST FOOD', 'FERRY/FERRY TERMINAL', 'FOOD SUPERMARKET', 'GAS STATION', 'GROCERY/BODEGA', 'GYM/FITNESS FACILITY', 'HIGHWAY/PARKWAY', 'HOMELESS SHELTER', 'HOSPITAL', 'HOTEL/MOTEL', 'JEWELRY', 'LIQUOR STORE', 'LOAN COMPANY', 'MAILBOX INSIDE', 'MAILBOX OUTSIDE', 'MARINA/PIER', 'MOSQUE', 'OPEN AREAS (OPEN LOTS)', 'OTHER', 'OTHER HOUSE OF WORSHIP', 'PARK/PLAYGROUND', 'PARKING LOT/GARAGE (PRIVATE)', 'PARKING LOT/GARAGE (PUBLIC)', 'PHOTO/COPY', 'PRIVATE/PAROCHIAL SCHOOL', 'PUBLIC BUILDING', 'PUBLIC SCHOOL', 'RESIDENCE - 

The values in `"PREM_TYP_DESC"` are clearly meant to distinguish different type of locations. Some are similar, and can potentially be merged into common categories if the need should arise (such as the mulitple type of residences).


In [411]:
drop_columns(df,["LOC_OF_OCCUR_DESC"])

21 columns before the drop.
20 columns after the drop.


## 3 . 4 . Special Locations

There are a few columns that provide special location information.
* `"HADEVELOPT"` (DROP)
    > Name of a New York City Housing Authority Development, if applicable.
* `"PARKS_NM"` (DROP)
    >  Name of a New York City park, playground, or greenspace, if applicable.
* `"STATION_NAME"` (DROP)
    > Name of a Transit Station, if applicable.
* `"TRANSIT_DISTRICT"` (DROP)
    > Name of a Transit District, if applicable.

Each of these features probably provides too much resolution or nuance for the purposes of this project. I can check that each of these columns is strongly associated with the corresponding descriptions from the `"PREM_TYP_DESC"` column.



In [412]:
print("Greatest Premise Type Description for Each Special Location Type")
for col in ["HADEVELOPT","PARKS_NM","STATION_NAME","TRANSIT_DISTRICT"]:
    # print the most frequent of the premise types that has a value in the iterated column
    print(f"{col:>16} : {df.loc[ ~df[col].isna(), 'PREM_TYP_DESC'].value_counts().index.to_list()[0]}")
    # print the number of occurences for the mose frequent of the premise types and all premise types that have values in the iterated column
    print(f"{'':>16} : {df.loc[ ~df[col].isna(), 'PREM_TYP_DESC'].value_counts()[0]} out of {df.loc[ ~df[col].isna(), 'PREM_TYP_DESC'].value_counts().sum()} data entries")
    # print the percent of occurences for the most frequent of the premise types that has a value in the iterated column
    print(f"{'':>16} : {df.loc[ ~df[col].isna(), 'PREM_TYP_DESC'].value_counts()[0]/df.loc[ ~df[col].isna(), 'PREM_TYP_DESC'].value_counts().sum()*100:.2f}%")

Greatest Premise Type Description for Each Special Location Type
      HADEVELOPT : RESIDENCE - PUBLIC HOUSING
                 : 1530 out of 1572 data entries
                 : 97.33%
        PARKS_NM : PARK/PLAYGROUND
                 : 2682 out of 3192 data entries
                 : 84.02%
    STATION_NAME : TRANSIT - NYC SUBWAY
                 : 8840 out of 8976 data entries
                 : 98.48%
TRANSIT_DISTRICT : TRANSIT - NYC SUBWAY
                 : 8840 out of 8976 data entries
                 : 98.48%


The correspondence is overwhelming, so I will drop the special locations from the dataframe.

In [413]:
drop_columns(df,["HADEVELOPT","PARKS_NM","STATION_NAME","TRANSIT_DISTRICT"])

20 columns before the drop.
16 columns after the drop.


## 3 . 5 . Jurisdiction

The jurisdiction for that is responsible for managing the response to incident can belong to one of 18 different departments. As it is unlikely that a significant number of incidences are carried out based on which authority's jurisdiction it will happen, I will drop these columns.
* `"JURISDICTION_CODE"` (DROP)
    > Numerical designation for a jurisdiction.
* `"JURIS_DESC"` (DROP)
    > Description name for a jurisdiction.



In [414]:
drop_columns(df,["JURISDICTION_CODE","JURIS_DESC"])

16 columns before the drop.
14 columns after the drop.


## 3 . 6 . Offense Codes

There are a few additional numerical codes that are used in this data set.

* `"OFNS_DESC"` (KEEP)
    >Description of the offense. Types of crimes may be relevant, as differnt types will likely have different distributions throughout the city.

* `"PD_DESC"` (DROP)
    >These codes more greatly refine the encoding of the `"OFNS_DESC"` column, so would indicate extra information about specific crimes which goes beyond identifying general types of crime.

* `"KY_CD"` (DROP)
    >These are numerical codes corresponding to the descriptive values of the `"OFNS_DESC"` column. This is redundent information for this analysis.

* `"PD_CD"` (DROP)
    >These codes more greatly refine the encoding of the `"KY_CD"` column, so would indicate extra information about specific crimes which goes beyond identifying general types of crime.


In [415]:
drop_columns(df,["PD_DESC","KY_CD","PD_CD"])

14 columns before the drop.
11 columns after the drop.


## 3 . 7 . Crime Completion

* `"CRM_ATPT_CPTD_CD"` (DROP)
    > Whether or not the crime was successfully completed or not doesn't alter the fact a crime was initiated.

In [416]:
drop_columns(df,["CRM_ATPT_CPTD_CD"])

11 columns before the drop.
10 columns after the drop.


## 3 . 8 . Additional Codes


* `"HOUSING_PSA"` (DROP) 
    >I am assuming this column is related to the Police Service Areas. There are nine Police Service Areas (PSA) in New York City, which patrol housing developments within 31 different precincts. Since this column has 362 unique values, I am not sure what it represents. Since each PSA contains mulitiple precincts, they are individually larger than the resolution I will be examining. I can only guess the many values in this column narrow down the location to a resolution smaller than what I want.


In [417]:
drop_columns(df,["HOUSING_PSA"])

10 columns before the drop.
9 columns after the drop.


## 3 . 9 . Time and Date

* `"CMPLNT_FR_DT"` and `"CMPLNT_FR_TM"` (KEEP)
    >When the event happens or begins is very relevant. All the data rows have these values present.
* `"CMPLNT_TO_DT"` and `"CMPLNT_TO_TM"` (DROP)
    >If present, these values indicate a time range when the crime occured. This column has many missing values as many crimes are known to happen at specific times. Since crime data is used to understand and to try to reduce crime, there is more value knowing when crimes are likely to begin rather than when they are likely to end. A crime can't be prevented if it has already happened.
* `"RPT_DT"` (DROP)
    > The date the crime was reported to the police does not alter when the crime actually occured.

In [418]:
drop_columns(df,["CMPLNT_TO_DT","CMPLNT_TO_TM","RPT_DT"])

9 columns before the drop.
6 columns after the drop.


# 4 . Data Types


Of the remaining columns, I will review the dataframe information.

In [419]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 449506 entries, 903695881 to 806537994
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   ADDR_PCT_CD    449506 non-null  int64 
 1   CMPLNT_FR_DT   449506 non-null  object
 2   CMPLNT_FR_TM   449506 non-null  object
 3   LAW_CAT_CD     449506 non-null  object
 4   OFNS_DESC      449497 non-null  object
 5   PREM_TYP_DESC  448236 non-null  object
dtypes: int64(1), object(5)
memory usage: 24.0+ MB


Here, the date and time columns are the object data type, but would probably be better served as the datetime datatype. However, I have encounted an error in trying to convert the string into datetime values, so I will need to examine the strings more carefully.

In [448]:
# Examine the month values
display(df["CMPLNT_FR_DT"].apply(lambda x: x[0:2]).value_counts().sort_index())

01    33767
02    28572
03    34956
04    32982
05    36990
06    37879
07    39779
08    39383
09    40196
10    43288
11    41615
12    40099
Name: CMPLNT_FR_DT, dtype: int64

All the month values are individually OK.

In [449]:
# Examine the day values
display(df["CMPLNT_FR_DT"].apply(lambda x: x[3:5]).value_counts().sort_index())

01    17941
02    14340
03    14349
04    14589
05    14877
06    14604
07    14440
08    14357
09    14441
10    15047
11    14703
12    15011
13    14925
14    14741
15    15713
16    15145
17    14989
18    14373
19    14618
20    15261
21    14915
22    14592
23    14648
24    14695
25    14483
26    14510
27    14435
28    14207
29    12944
30    13552
31     8061
Name: CMPLNT_FR_DT, dtype: int64

All the day values are individually OK. (The days 29, 30, and 31 might not work with some month combinations if those months do not have those days.)

In [451]:
# Examine the year values
display(df["CMPLNT_FR_DT"].apply(lambda x: x[6:]).value_counts().sort_index())

1019         1
1021         7
1920         1
1921         8
1955         1
1958         1
1960         1
1966         1
1967         1
1969         1
1970         1
1971         7
1972         1
1973         1
1976         2
1977         1
1978         1
1979         3
1980         1
1981         4
1982         2
1983         3
1984         5
1985         1
1986         3
1987         4
1988         5
1989         5
1990         5
1991         3
1992         1
1993         4
1994         5
1995         8
1996         4
1997        11
1998        10
1999        10
2000        18
2001        29
2002        14
2003        16
2004        13
2005        16
2006        23
2007        32
2008        38
2009        40
2010        46
2011        61
2012        99
2013        94
2014       111
2015       116
2016       175
2017       244
2018       335
2019       898
2020      7329
2021    439625
Name: CMPLNT_FR_DT, dtype: int64

Well, one row has the year 1019 and seven rows have the year 1021. Those are certaintly wrong.

Additionally, the number of reports for all years before 2021 are significantly fewer. So I will just drop all the rows with a year that isn't 2021.

In [456]:
filter_2021 = df["CMPLNT_FR_DT"].apply(lambda x: x[6:])=="2021"
df = df.loc[filter_2021, :]

Now the unnecessary year (and invalid years) are gone, I can convert the strings to the datetime data type.

In [459]:
df["CMPLNT_FR_DT"] = pd.to_datetime(df["CMPLNT_FR_DT"])

In [462]:
df["CMPLNT_FR_TM"] = pd.to_datetime(df["CMPLNT_FR_TM"])

In [463]:
df["CMPLNT_FR_TM"]

CMPLNT_NUM
903695881   2022-04-15 22:13:00
400462399   2022-04-15 06:21:00
587910690   2022-04-15 20:05:00
186105368   2022-04-15 22:49:00
185325394   2022-04-15 17:25:00
                    ...        
702183776   2022-04-15 15:40:00
798000815   2022-04-15 07:00:00
900510890   2022-04-15 04:50:00
244132985   2022-04-15 09:35:00
366756177   2022-04-15 19:40:00
Name: CMPLNT_FR_TM, Length: 439625, dtype: datetime64[ns]

In [464]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 439625 entries, 903695881 to 366756177
Data columns (total 6 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   ADDR_PCT_CD    439625 non-null  int64         
 1   CMPLNT_FR_DT   439625 non-null  datetime64[ns]
 2   CMPLNT_FR_TM   439625 non-null  datetime64[ns]
 3   LAW_CAT_CD     439625 non-null  object        
 4   OFNS_DESC      439617 non-null  object        
 5   PREM_TYP_DESC  438418 non-null  object        
dtypes: datetime64[ns](2), int64(1), object(3)
memory usage: 23.5+ MB


In [423]:
for col in df.columns:
    print(f"{col:>17} : {df[col].nunique():>4} : {df[col].value_counts().index.to_list()}")

      ADDR_PCT_CD :   77 : [75, 43, 114, 47, 44, 40, 14, 46, 52, 109, 67, 73, 42, 115, 110, 48, 105, 103, 19, 113, 79, 18, 70, 120, 104, 25, 13, 83, 1, 32, 106, 71, 49, 62, 108, 23, 102, 121, 90, 60, 9, 45, 34, 7, 33, 107, 28, 77, 72, 41, 6, 61, 81, 84, 122, 50, 101, 24, 68, 63, 69, 30, 66, 20, 94, 112, 78, 10, 88, 5, 17, 111, 26, 100, 123, 76, 22]
     CMPLNT_FR_DT : 1979 : ['01/01/2021', '10/15/2021', '10/01/2021', '12/01/2021', '10/20/2021', '09/17/2021', '11/15/2021', '10/08/2021', '11/19/2021', '10/21/2021', '06/01/2021', '10/22/2021', '10/13/2021', '11/12/2021', '10/14/2021', '09/23/2021', '09/24/2021', '11/05/2021', '12/02/2021', '12/03/2021', '09/27/2021', '10/06/2021', '11/18/2021', '11/01/2021', '09/15/2021', '09/22/2021', '11/16/2021', '11/10/2021', '11/09/2021', '10/16/2021', '09/25/2021', '10/07/2021', '07/15/2021', '11/20/2021', '10/12/2021', '09/16/2021', '11/30/2021', '11/17/2021', '10/28/2021', '12/10/2021', '12/04/2021', '07/16/2021', '09/28/2021', '07/23/2021', '11/1

In [424]:
show_column_values(df,df.columns)

       ADDR_PCT_CD :    77 : [1, 5, 6, 7, 9, 10, 13, 14, 17, 18, 19, 20, 22, 23, 24, 25, 26, 28, 30, 32, 33, 34, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 52, 60, 61, 62, 63, 66, 67, 68, 69, 70, 71, 72, 73, 75, 76, 77, 78, 79, 81, 83, 84, 88, 90, 94, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 120, 121, 122, 123]
                   :     0 : Rows Missing Data

      CMPLNT_FR_DT :  1979 : ['01/01/1955', '01/01/1967', '01/01/1972', '01/01/1976', '01/01/1978', '01/01/1979', '01/01/1980', '01/01/1981', '01/01/1983', '01/01/1984', '01/01/1985', '01/01/1986', '01/01/1987', '01/01/1988', '01/01/1989', '01/01/1990', '01/01/1991', '01/01/1993', '01/01/1994', '01/01/1995', '01/01/1996', '01/01/1997', '01/01/1998', '01/01/1999', '01/01/2000', '01/01/2001', '01/01/2002', '01/01/2003', '01/01/2004', '01/01/2005', '01/01/2006', '01/01/2007', '01/01/2008', '01/01/2009', '01/01/2010', '01/01/2011', '01/01/2012', '01/01/2013', '01/01/2014', '01/01/2015', '01/01/20